kernel/src/expressions/column_name.rs

```diff
@@ -0,0 +1,313 @@
+/// use [`nested_column_name!`] instead.
+// NOTE: Macros are only public if exported, which defines them at the root of the crate. But we
+// don't want it there. So, we export a hidden macro and pub use it here where we actually want it.
+#[macro_export]
+#[doc(hidden)]
+macro_rules! __simple_column_name {
+    ( $($name:tt)* ) => {
+        $crate::expressions::ColumnName::new(delta_kernel_derive::parse_simple_column_name!($($name)*))
+    };
+}
+#[doc(inline)]
+pub use __simple_column_name as simple_column_name;
+
+/// Creates a nested column name whose field names are all simple column names. This macro is
+/// provided as a convenience for the common case where the caller knows the column contains only
+/// simple field names and that splitting by periods is safe.
+///
+/// To avoid accidental misuse, the argument must be a string literal. Thus, the following
+/// invocation would fail to compile:
+///
+/// ```fail_compile
+/// # use delta_kernel::expressions::nested_column_name;
+/// let s = "a.b";
+/// let name = nested_column_name!(s);
+/// ```
+///
+/// NOTE: `nested_column_name!("a.b.c")` produces a path with field names "a", "b", and "c".
+#[macro_export]
+#[doc(hidden)]
+macro_rules! __nested_column_name {
+    ( $($name:tt)* ) => {
+        $crate::expressions::ColumnName::new(delta_kernel_derive::parse_nested_column_name!($($name)*))
+    };
+}
+#[doc(inline)]
+pub use __nested_column_name as nested_column_name;
+
+/// Joins two column names together, when one or both inputs might be literal strings representing
+/// simple (non-nested) column names. For example, the invocation `joined_column_name!("a", "b")` is
+/// equivalent to `ColumnName::join(simple_column_name!("a"), simple_column_name!("b"))`
+///
+/// To avoid accidental misuse, at least one argument must be a string literal. Thus, the following
+/// invocation would fail to compile:
+///
+/// ```fail_compile
+/// # use delta_kernel::expressions::joined_column_name;
+/// let s = "s";
+/// let name = joined_column_name!(s, s);
+/// ```
+///
+/// NOTE: `joined_column_name!("a.b", "c")` produces a path with field names "a.b" and "c". If
+/// either argument is a nested path delimited by periods, it should be split first and the result
+/// passed to `joined_column_name!`, e.g. `joined_column_name!(nested_column_name!("a.b"), "c")`.
+#[macro_export]
+#[doc(hidden)]
+macro_rules! __joined_column_name {
+    ( $left:literal, $right:literal ) => {
+        $crate::expressions::ColumnName::join(
+            $crate::__simple_column_name!($left),
+            $crate::__simple_column_name!($right),
+        )
+    };
+    ( $left:literal, $right:expr ) => {
+        $crate::expressions::ColumnName::join($crate::__simple_column_name!($left), $right)
+    };
+    ( $left:expr, $right:literal) => {
+        $crate::expressions::ColumnName::join($left, $crate::__simple_column_name!($right))
+    };
```

Comment:
With the above suggestion this becomes:

```suggestion
    ( $left:literal, $right:literal ) => {
        $crate::__simple_column_name!($left).join($crate::__simple_column_name!($right))
    };
    ( $left:literal, $right:expr ) => {
        $crate::__simple_column_name!($left).join($right)
    };
    ( $left:expr, $right:literal) => {
        $left.join($crate::__simple_column_name!($right))
    };
```